File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/interfaces/BillingAddressAttributes.java
package com.emerchantpay.gateway.genesisandroid.api.interfaces;

import com.emerchantpay.gateway.genesisandroid.api.models.Country;
import com.emerchantpay.gateway.genesisandroid.api.util.RequestBuilder;

public interface BillingAddressAttributes {

    RequestBuilder requestBuilder = new RequestBuilder("");

    // Billing PaymentAddress
    default BillingAddressAttributes setBillingFirstname(String firstname) {
        requestBuilder.addElement("first_name", firstname);
        return this;
    }

    default BillingAddressAttributes setBillingLastname(String lastname) {
        requestBuilder.addElement("last_name", lastname);
        return this;
    }

    default BillingAddressAttributes setBillingPrimaryAddress(String primaryAddress) {
        requestBuilder.addElement("address1", primaryAddress);
        return this;
    }

    default BillingAddressAttributes setBillingSecondaryAddress(String secondaryAddress) {
        requestBuilder.addElement("address2", secondaryAddress);
        return this;
    }

    default BillingAddressAttributes setBillingCity(String city) {
        requestBuilder.addElement("city", city);
        return this;
    }

    default BillingAddressAttributes setBillingZipCode(String zipCode) {
        requestBuilder.addElement("zip_code", zipCode);
        return this;
    }

    default BillingAddressAttributes setBillingState(String state) {
        requestBuilder.addElement("state", state);
        return this;
    }

    default BillingAddressAttributes setBillingCountry(String country) throws IllegalAccessException {
        Country c = new Country();

        requestBuilder.addElement("country", c.getIsoCode(country));

        return this;
    }

    default RequestBuilder buildBillingAddress() {
        return requestBuilder;
    }
}

File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/interfaces/BillingAddressAttributes.kt
package com.emerchantpay.gateway.genesisandroid.api.interfaces

import com.emerchantpay.gateway.genesisandroid.api.models.Country
import com.emerchantpay.gateway.genesisandroid.api.util.RequestBuilder


interface BillingAddressAttributes {

    // Billing PaymentAddress
    fun setBillingFirstname(firstname: String): BillingAddressAttributes {
        requestBuilder.addElement("first_name", firstname)
        return this
    }

    fun setBillingLastname(lastname: String): BillingAddressAttributes {
        requestBuilder.addElement("last_name", lastname)
        return this
    }

    fun setBillingPrimaryAddress(primaryAddress: String): BillingAddressAttributes {
        requestBuilder.addElement("address1", primaryAddress)
        return this
    }

    fun setBillingSecondaryAddress(secondaryAddress: String): BillingAddressAttributes {
        requestBuilder.addElement("address2", secondaryAddress)
        return this
    }

    fun setBillingCity(city: String): BillingAddressAttributes {
        requestBuilder.addElement("city", city)
        return this
    }

    fun setBillingZipCode(zipCode: String): BillingAddressAttributes {
        requestBuilder.addElement("zip_code", zipCode)
        return this
    }

    fun setBillingState(state: String): BillingAddressAttributes {
        requestBuilder.addElement("state", state)
        return this
    }

    @Throws(IllegalAccessException::class)
    fun setBillingCountry(country: String): BillingAddressAttributes {
        val c = Country()

        requestBuilder.addElement("country", c.getIsoCode(country)!!)

        return this
    }

    fun buildBillingAddress(): RequestBuilder {
        return requestBuilder
    }

    companion object {
        val requestBuilder = RequestBuilder("")
    }
}